File: 027_Capstone_Project/Capstone_Consumer.py
from pyspark.streaming.kafka import KafkaUtils
from pyspark.streaming import StreamingContext
from pyspark import SparkContext
from pyspark.conf import SparkConf
from pyspark.sql import SparkSession
from pyspark import sql
from pyspark.sql import SQLContext
from pyspark.sql import types
import json
import csv
from json import loads
from time import sleep



print("PROGRAM START!!!")
print("PROGRAM START!!!")
print("PROGRAM START!!!")
print("PROGRAM START!!!")

sc= SparkContext()
ssc = StreamingContext(sc, 10)
sqlc= SQLContext(sc)
directKafkaStream = KafkaUtils.createDirectStream(ssc, ["Lufthansa"], {"metadata.broker.list": "localhost:9195"})
lines= directKafkaStream.map(lambda x: x[1])

print("LINES START!!!")
print("LINES START!!!")
print("LINES START!!!")
print("LINES START!!!")

def transformer(rdd):
	my_obj= json.loads(rdd)
	return (my_obj["Departure"]["AirportCode"],my_obj["Departure"]["ScheduledTimeLocal"]["DateTime"],my_obj["Departure"]["ScheduledTimeUTC"]["DateTime"],my_obj["Departure"]["TimeStatus"]["Code"],my_obj["Arrival"]["AirportCode"],my_obj["Arrival"]["ScheduledTimeLocal"]["DateTime"],my_obj["Arrival"]["ScheduledTimeUTC"]["DateTime"],my_obj["Arrival"]["TimeStatus"]["Code"],my_obj["OperatingCarrier"]["AirlineID"],my_obj["OperatingCarrier"]["FlightNumber"],my_obj["FlightStatus"]["Definition"],my_obj["ServiceType"])

transform= lines.map(transformer)
print(transform)

def build_df(rdd):
	if not rdd.isEmpty():
		global sqlc
		df= sqlc.createDataFrame(rdd, schema= ["Departure","D/Time(Local.)","D/Time(UTC)","Status.","Arrival","D/Time(Local)","D/Time(UTC.)","Status","Airline_ID","Flight_No","Flight_Status","Service"])
		df.write.format('jdbc').options(url='jdbc:mysql://localhost/BigData',driver='com.mysql.jdbc.Driver',dbtable='Lufthansa',user='fieldemployee',password='Password').mode('append').save()
		df.show()

transform.foreachRDD(build_df)


ssc.start()
ssc.awaitTermination()

File: 027_Capstone_Project/Capstone_Consumer.scala
import org.apache.spark.streaming.StreamingContext._
import org.apache.spark._
import org.apache.spark.streaming._
import org.apache.spark.sql._
import org.apache.spark.sql.types._
import org.apache.spark.streaming.kafka._
import java.util.Properties
import org.apache.spark.sql.SparkSession
import org.apache.spark.SparkConf
import scala.collection.mutable.ListBuffer

object consumer
{	
	def main(args: Array[String]): Unit =
	{
		val linelist: ListBuffer[String] = new ListBuffer[String]();
		val conf = new SparkConf()
		val ss = SparkSession.builder.master("local").appName("Lufthansa").config(conf).getOrCreate()
		import ss.implicits._
		val sc = ss.sparkContext
 		val ssc = new StreamingContext(sc, Seconds(5))
		val sqlContext = new org.apache.spark.sql.SQLContext(sc)
		val dstream = KafkaUtils.createStream(ssc, "localhost:9195", "mygroup" , Map("Lufthansa"-> 1))
		val prop = new Properties()
		prop.setProperty("user", "fieldemployee")
		prop.setProperty("password", "Password")
		val schema = new StructType()
  		.add(StructField("Departure", StringType, false))
  		.add(StructField("D/Time(Local.)", StringType, false))
  		.add(StructField("D/Time(UTC)", StringType, false))
		.add(StructField("Status.", StringType, false))
		.add(StructField("Arrival", StringType, false))
		.add(StructField("D/Time(Local.)", StringType, false))
		.add(StructField("D/Time(UTC.)", StringType, false))
		.add(StructField("Status", StringType, false))
		.add(StructField("Airline_ID", StringType, false))
		.add(StructField("Flight_No", StringType, false))
		.add(StructField("Flight_Status", StringType, false))
		.add(StructField("Service", StringType, false))
		val mappedlines = dstream.foreachRDD{ rdd =>
      		rdd.foreach{record =>
			     var mylist = List(record._2)
			     print(mylist)
			     var myrdd = sc.parallelize(mylist)		     
			     var myrdd2 = myrdd.map{x => Row(x(0),x(1),x(2),x(3),x(4),x(5),x(6),x(7),x(8),x(9),x(10),x(11))}
			     var df = sqlContext.createDataFrame(myrdd2, schema)
			     // Create DF using createDataframe
			     if (df != null)
			     {
				df.write.jdbc(s"jdbc:mysql://localhost:3306/BigData", "Lufthansa", prop)
			     }
			   }
		}
		ssc.start
   		ssc.awaitTermination
	}
}
